crucible/src/Lang/Crucible/CFG/Expr.hs

```diff
@@ -1360,3 +1399,207 @@ foldApp f0 r0 a = execState (traverseApp (go f0) a) r0
 mapApp :: TraversableFC (ExprExtension ext)
        => (forall u . f u -> g u) -> App ext f tp -> App ext g tp
 mapApp f a = runIdentity (traverseApp (pure . f) a)
+
+-------------------------------------------------------------------------------------
+-- Type Instantiation
+
+-- App :: Type -> (CrucibleType -> Type) -> CrucibleType -> Type
+type instance Instantiate subst App = App  
+instance (IsSyntaxExtension ext) => InstantiateFC (App ext) where
```

Comment:
Here, in particular, I think a traversal would eliminate most of this boilerplate.

Reply:
I heartily agree. But note: there are a few special cases for reasoning about closed subcomponents.